controllers/lookup/k8s_client.go

```diff
@@ -0,0 +1,160 @@
+package lookup
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/datainfrahq/druid-operator/apis/druid/v1alpha1"
+	internalhttp "github.com/datainfrahq/druid-operator/pkg/http"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"net/http"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type K8sClient struct {
+	client client.Client
+}
+
+func NewK8sClient(client client.Client) K8sClient {
+	return K8sClient{
+		client: client,
+	}
+}
+
+func (c *K8sClient) FindLookups(ctx context.Context, reports map[types.NamespacedName]Report) (map[ClusterKey]map[LookupKey]Spec, error) {
+	specs := &v1alpha1.DruidLookupList{}
+	if err := c.client.List(ctx, specs); err != nil {
+		return nil, err
+	}
+
+	lookupSpecsPerCluster := make(map[ClusterKey]map[LookupKey]Spec)
+	for _, spec := range specs.Items {
+		clusterKey := ClusterKey{
+			Namespace: spec.Namespace,
+			Cluster:   spec.Spec.DruidClusterName,
+		}
+		lookupKey := LookupKey{
+			Tier: spec.Spec.Tier,
+			Id:   spec.Spec.Id,
+		}
+		var lookupSpec interface{}
+		if err := json.Unmarshal([]byte(spec.Spec.Spec), &lookupSpec); err != nil {
+			setIfNotPresent(
+				reports,
+				spec.GetNamespacedName(),
+				Report(NewErrorReport(
+					fmt.Errorf(
+						"lookup resource %v in cluster %v/%v contains invalid spec, should be JSON",
+						spec.Name,
+						clusterKey.Namespace,
+						clusterKey.Cluster,
+					),
+				)),
+			)
+			continue
+		}
+
+		if lookupKey.Tier == "" {
+			lookupKey.Tier = "__default"
+		}
+
+		if lookupSpecsPerCluster[clusterKey] == nil {
+			lookupSpecsPerCluster[clusterKey] = make(map[LookupKey]Spec)
+		}
+		ls := lookupSpecsPerCluster[clusterKey]
+
+		if _, replaced := replace(ls, lookupKey, Spec{
+			name: spec.GetNamespacedName(),
+			spec: lookupSpec,
+		}); replaced {
```

Comment:
Can you explain what is the logic you're trying to achieve here?

Reply:
I tend to prefer a coding style where I avoid separating a check and a action if the result of the check can't be expressed in the type system. I find it helpful to avoid separating the place the performs a check and the place that relies on that check and thus running into some edge case.

In this case (and the other instances where I've used `replace`) it could be replaced by

```
if _, ok := ls[lookupKey] {
    // diverging if body
}

ls[lookupKey] = Spec{
    name: spec.GetNamespacedName(),
    spec: lookupSpec,
}
```

If this style is preferred in this project, I'd be happy to convert this occurrence and the others.

Reply:
Addressed in abbe6ff

List and map functions inlined.